### PR TITLE
Add support for optional chain in assignments

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2824 (partial)
| Any Dependency Changes?  |
| License                  | MIT

Implementation of https://github.com/nicolo-ribaudo/proposal-optional-chaining-assignment, that reached stage 1 at the July 2023 TC39 meeting. During the meeting there was some debate about syntax (i.e. replacing `=` with another token), so for now I'm adding a required `version: "2023-07"` option that will allow us to introduce breaking changes in the future.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15751"></a>

### CODE REVIEW

File: packages/babel-plugin-proposal-optional-chaining-assign/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# @babel/plugin-transform-optional-chaining
```
Comment (CanadaHonk) [inline packages/babel-plugin-proposal-optional-chaining-assign/README.md:1]: Fyi, this README is misnamed and should be `@babel/plugin-proposal-optional-chaining-assign`.

File: packages/babel-plugin-proposal-optional-chaining-assign/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# @babel/plugin-proposal-optional-chaining-assign
+
+> Transform optional chaining on the left-hand side of assignment expressions
+
+See our website [@babel/plugin-proposal-optional-chaining-assign](https://babeljs.io/docs/babel-plugin-proposal-optional-chaining-assign) for more information.
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-optional-chaining-assign/README.md:5]: Can you create a docs PR for this package?

**Review by JLHwung — COMMENTED** [review]
The plugin looks good to me generally. Can you add test cases for private property, e.g.
```js
a?.#b = c
```
with and without the class properties transform? I recall that we have to handle optional chain in the `packages/babel-helper-member-expression-to-functions`, I guess we should handle the assignment as well, since the class field transform is invoked earlier than the expression.

File: packages/babel-helper-member-expression-to-functions/src/index.ts [hunk 3]
```diff
@@ -432,44 +435,7 @@ const handle = {
     // MEMBER += VALUE   ->   _set(MEMBER, _get(MEMBER) + VALUE)
     // MEMBER ??= VALUE   ->   _get(MEMBER) ?? _set(MEMBER, VALUE)
     if (parentPath.isAssignmentExpression({ left: node })) {
-      if (this.simpleSet) {
```
Comment (nicolo-ribaudo) [inline packages/babel-helper-member-expression-to-functions/src/index.ts:435]: All this section is moved to the `handleAssignment` function below with no changes (other than renaming `this` to `state`).

**Review by nicolo-ribaudo — COMMENTED** [review]
@JLHwung Thanks, I fixed `a?.#b = c` support and added tests.

**Review by JLHwung — APPROVED** [review]

**Review by existentialism — APPROVED** [review]
❤️

**Pull request merged**